### ISSUE TITLE
Add type hints from typeshed

### ISSUE THREAD

**Issue by hugovk** [jazzband/prettytable#203]
Let's add type hints to PrettyTable.

The good news is much of the work is already done because typeshed has .pyi signature stubs:

* https://github.com/python/typeshed/tree/master/stubs/prettytable/prettytable

And here are some resources for automating the conversion of stubs into inline annotations:

* https://typing.readthedocs.io/en/latest/#type-hint-and-stub-integration

Tasks:

* [x] Add this to `.pre-commit-config.yaml` so pre-commit adds this import to Python files and we can use newer typing Syntax on older Python 3.7:

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 3dbc9c0..316e7f5 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -15,6 +15,7 @@ repos:
     rev: 5.10.1
     hooks:
       - id: isort
+        args: [--add-import=from __future__ import annotations]
 
   - repo: https://github.com/PyCQA/flake8
     rev: 5.0.4
```

* [x] Apply the type hints from typeshed to this repo
* [ ] Add mypy via pre-commit and make mypy pass (can be follow-up PR if there's much to do)
* [x] Add py.typed and "Typing :: Typed" Trove classifier

**Comment by phershbe** [comment]
I can work on this if you want. I'm relatively new but I understand type hinting. I know the basics of how to run Mypy against type hints, but I am working on understanding Typeshed for this issue. Obviously the links explain how to kind of automate it with the .pyi signature stubs so that I don't have to type out the hint for each specific case ... I'm reading the documentation from the links now. I can work hard on it and hopefully be done really quickly.

**Comment by hugovk** [comment]
Sounds good, thank you! I've not used those tools either, but feel free to ask if you've any questions and we'll work it out :)

**Comment by phershbe** [comment]
@hugovk I'm working on this still. I didn't even know what stubs or .pyi files were until a few days ago; some PEPs explain them and your links above do too. I'm here to help the project not hinder progress though, so if you or anybody else want to submit the PR, go for it. I'm still here and working on it too!

**Comment by hugovk** [comment]
Thank you, if you want to keep working on it that's fine, no rush!

**Comment by phershbe** [comment]
@hugovk Since I'm relatively inexperienced and have never worked as a programmer before, stuff like this can take me a while to get my head around. I understand what typeshed is and what .pyi files are now, it's simple but it was confusing to me for a while. I'm working on doing this with merge_pyi. I should be able to get in the PR by tomorrow, I'll either get it in or post a message telling what's going on. If anybody else needs to finish it before then, like I said before, that's fine too. Thank you for your patience!

**Comment by hugovk** [comment]
Sounds good, we're in no hurry with this. Feel free to open a draft PR if you have questions or would like some feedback :)

**Comment by phershbe** [comment]
@hugovk Okay, thank you. I was using merge_pyi (https://github.com/google/merge_pyi), but 
`python3 merge_pyi.py --pep484 --diff prettytable.py prettytable.pyi`
gives me the error:
usage: merge_pyi.py [-h] [-w] [--as-comments] [--diff] file.py file.pyi
merge_pyi.py: error: unrecognized arguments: --pep484

`python3 merge_pyi.py prettytable.py prettytable.pyi` works but spits out the code in the terminal, and then the spaces separating comments are gone, so I'm trying to figure out how to get it to put the output in a file.

So I'm trying to figure all of this out. I'll get it, but I thought I would let you know where I'm at. And obviously it would be quicker at this point to take the annotations from typeshed and put them in myself, but I'm being stubborn about figuring out how to do it automatically.

**Comment by hugovk** [comment]
Let's check the help using the `-h` or `--help` switch:

```console
$ python merge_pyi.py -h
usage: merge_pyi.py [-h] [-w] [--as-comments] [--diff] file.py file.pyi

Copy type annotations from file.pyi to file.py.

positional arguments:
  file.py        python file to annotate
  file.pyi       PEP484 stub file with annotations for file.py

options:
  -h, --help     show this help message and exit
  -w             overwrite file.py
  --as-comments  insert type annotations as comments
  --diff         print out a diff

Outputs merged file to stdout.
```

Try using the `-w` option :)

**Comment by phershbe** [comment]
@hugovk Awesome, very helpful! I submitted the draft PR, you can let me know what you think.

**Comment by hugovk** [comment]
I'd like to put out a release before the month is out, there's a couple of little things we need to do first (the pre-commit/mypy stuff can wait, that's a bigger job).

1. Create an empty `py.typed` file in `src/prettytable/`
   * This is a special "marker file" which indicates the package supports typing (all the gory details are in [PEP 561](https://peps.python.org/pep-0561/))

2. Add "Typing :: Typed" (alphabetically) to the `classifiers` list in `setup.cfg`
   * This is called a "[Trove classifier](https://pypi.org/classifiers/)" which is a keyword to help find and classify things on [PyPI](https://pypi.org/search/?q=&o=&c=Typing+%3A%3A+Typed)

@phershbe Would you like to take care of these?

Once that is done and released, some automation at https://github.com/python/typeshed will kick off to archive the old stubs there, and people can use them directly in this package!

**Comment by phershbe** [comment]
@hugovk Yeah I can do that, I have some free time now so I'll create the draft. Thank you for offering it to me!

As for the pre-commit and mypy stuff, I got my head wrapped around pre-commit and I got mypy added to the yaml file, but then the mypy raises 41 errors and I'm trying to work through those.

**Comment by hugovk** [comment]
Great!

Yes, there will be quite a few errors! Feel free to open a draft PR and we can work through them together, they're not always obvious :)

**Comment by hugovk** [comment]
Thank you, your work is now released in PrettyTable 3.5.0! 🚀 

* https://github.com/jazzband/prettytable/releases/tag/3.5.0
* https://pypi.org/project/prettytable/3.5.0/

**Comment by AlexWaygood** [comment]
> Once that is done and released, some automation at https://github.com/python/typeshed will kick off to archive the old stubs there, and people can use them directly in this package!

The typeshed automation has successfully kicked off! Thanks @hugovk and @phershbe 😀

- https://github.com/python/typeshed/pull/9023

**Comment by hugovk** [comment]
This issue can be closed now PRs https://github.com/jazzband/prettytable/pull/205, https://github.com/jazzband/prettytable/pull/211 and https://github.com/python/typeshed/pull/9023 are merged 🚀 

https://github.com/jazzband/prettytable/issues/206 will be fixed by PR https://github.com/jazzband/prettytable/pull/218.

**Comment by hugovk** [comment]
The old stubs have now been removed from typeshed! :tada: https://github.com/python/typeshed/issues/9028 https://github.com/python/typeshed/pull/9988
